pkg/core/consensus/generation/broker.go

```diff
@@ -100,40 +117,55 @@ func (b *broker) Listen() {
 	for {
 		select {
 		case bid := <-b.bidChan:
-			b.proofGenerator.UpdateBidList(bid)
+			if b.proofGenerator != nil {
+				b.proofGenerator.UpdateBidList(bid)
+			}
 		case state := <-b.regenerationChan:
-			if state.Round == b.seeder.Round() {
+			if state.Round == b.seeder.Round() && b.proofGenerator != nil {
 				b.forwarder.threshold.Lower()
 				b.generateProofAndBlock()
 			}
 		case winningBlockHash := <-b.winningBlockHashChan:
 			cert := b.certificateGenerator.generateCertificate()
 			b.sendCertificateMsg(cert, winningBlockHash)
+		case blk := <-b.acceptedBlockChan:
+			if b.proofGenerator != nil {
+				b.onBlock(blk)
+			}
 		}
 	}
 }
 
-func (b *broker) onBlock(m *bytes.Buffer) error {
+func (b *broker) onBlock(blk block.Block) error {
 	b.forwarder.threshold.Reset()
 
-	blk := block.NewBlock()
-	if err := blk.Decode(m); err != nil {
-		return err
-	}
-
 	// Remove old bids before generating a new score
 	b.proofGenerator.RemoveExpiredBids(blk.Header.Height + 1)
 
 	return b.handleBlock(blk)
 }
 
-func (b *broker) handleBlock(blk *block.Block) error {
+func (b *broker) handleBlock(blk block.Block) error {
 	if err := b.seeder.GenerateSeed(blk.Header.Height+1, blk.Header.Seed); err != nil {
 		return err
 	}
 
-	b.forwarder.setPrevBlock(*blk)
-	b.generateProofAndBlock()
+	b.forwarder.setPrevBlock(blk)
+
+	// Only generate if we are allowed to
+	if b.proofGenerator.InBidList() {
+		b.generateProofAndBlock()
+	} else {
+		// We will remove the old proofgenerator, to avoid creation of obsolete proofs
+		// until we get new proof values
+		b.proofGenerator = nil
+
+		// Then, we will wait till a new bid belonging to us is found
+		go func() {
+			b.updateProofValues()
```

Comment:
The goroutine is only spawned once, as the `proofGenerator` is set to nil and `onBlock` can no longer be triggered if the `proofGenerator` is nil. It should exist until the component finds a new valid bid, or until the node is shut off.

Reply:
Not sure why a simple proofGenerator bool flag was not used instead of resetting the instance but anyway.

Reply:
Because the `proofGenerator` is an interface on the `broker`, and we can't change it's state directly. Unless we added a method for it of course, but for now I went for the simpler option.